### PR TITLE
Handle readonly property backing field logic during deserialization instead of serialization

### PR DESCRIPTION
**Pull request by justing-bq**
The ordinary deserialization logic does not work for readonly properties. To get around this we had been serializing the backing fields for readonly properties so we can still set the underlying value of the property during deserialization. However this results in those backing fields being inserted (as an extra column) into QLDB whenever the driver does an insert with an object with a readonly property.

ie. 
image

Now the readonly property / backing field logic is handled exclusively during deserialization instead of serialization (so the backing field logic is completely hidden from the end users).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Amazon.IonObjectMapper/IonObjectSerializer.cs [hunk 1]
```diff
@@ -273,12 +257,31 @@ private bool TryDeserializeField(FieldInfo field, IIonReader reader, IonType ion
...
+        }
+        
+        private bool IgnoreDeserializedField(FieldInfo field, object deserialized)
+        {
+            // Check if this field is really a backing field for a readonly property and
+            // if so, apply the ignore property logic instead of the ignore field logic.
+            if (!options.IgnoreReadOnlyProperties && IsBackingFieldForReadonlyProperty(field))
             {
-                return false;
+                return IgnoreDeserializedProperty(deserialized);
```
Comment (byronlin13) [inline Amazon.IonObjectMapper/IonObjectSerializer.cs:274]: Is this necessary? L278 has the same logic.

Comment (justing-bq) [reply Amazon.IonObjectMapper/IonObjectSerializer.cs:274]: `IgnoreDeserializedProperty` and `IgnoreDeserializedField` have similar but not identical logic.

Since this backing field is essentially a surrogate for a readonly property, we want to apply the property logic to this field (instead of the field logic).

File: Amazon.IonObjectMapper/IonObjectSerializer.cs [hunk 2]
```diff
@@ -173,11 +170,6 @@ public override void Serialize(IIonWriter writer, object item)
                     // This Ion property name was already serialized.
                     continue;
                 }
-                
-                if (property.GetCustomAttributes(true).Any(it => it is IonIgnore))
```
Comment (byronlin13) [inline Amazon.IonObjectMapper/IonObjectSerializer.cs:177]: How come we are removing this?

File: Amazon.IonObjectMapper/IonObjectSerializer.cs [hunk 3]
```diff
@@ -173,11 +170,6 @@ public override void Serialize(IIonWriter writer, object item)
...
-                {
-                    continue;
-                }
```
Comment (simonz-bq) [inline Amazon.IonObjectMapper/IonObjectSerializer.cs:180]: Why are we now serializing things with IonIgnore attribte?

File: Amazon.IonObjectMapper/IonObjectSerializer.cs (hunk 2 above)
Comment (justing-bq) [reply Amazon.IonObjectMapper/IonObjectSerializer.cs:177]: This check was removed because it was made redundant by the change made to `GetValidProperties`.

File: Amazon.IonObjectMapper/IonObjectSerializer.cs (hunk 3 above)
Comment (justing-bq) [reply Amazon.IonObjectMapper/IonObjectSerializer.cs:180]: This check was removed because it was made redundant by the change made to `GetValidProperties`.

File: Amazon.IonObjectMapper/IonObjectSerializer.cs [hunk 4]
```diff
@@ -273,12 +257,31 @@ private bool TryDeserializeField(FieldInfo field, IIonReader reader, IonType ion
...
+            
+            return (options.IgnoreReadOnlyFields && field.IsInitOnly) || 
+                   (options.IgnoreDefaults && deserialized == default);
+        }
 
-            return !options.IgnoreDefaults || deserialized != default;
+        private bool IsBackingFieldForReadonlyProperty(FieldInfo field)
+        {
+            var readonlyProperties = this.GetValidProperties(true).Where(IsReadOnlyProperty);
+            return readonlyProperties.Any(p => field.Name == $"<{p.Name}>k__BackingField");
```
Comment (simonz-bq) [inline Amazon.IonObjectMapper/IonObjectSerializer.cs:284]: Is string matching the only way to do this?

Comment (justing-bq) [reply Amazon.IonObjectMapper/IonObjectSerializer.cs:284]: This name comparison is the only way I know of to determine that a given field is a backing field for a given property.

File: Amazon.IonObjectMapper/IonObjectSerializer.cs (hunk 1 above)
Comment (byronlin13) [reply Amazon.IonObjectMapper/IonObjectSerializer.cs:274]: A backing field will have `field.IsInitOnly`= false because backing field only has `get`. It will go to L278 and it has the same conditional.

File: Amazon.IonObjectMapper/IonObjectSerializer.cs [hunk 5]
```diff
@@ -367,17 +370,34 @@ private PropertyInfo FindProperty(string readName)
...
+        private static bool IsReadOnlyProperty(PropertyInfo property)
         {
             return property.SetMethod == null;
         }
         
         private FieldInfo FindField(string name)
         {
             var exact = targetType.GetField(name, BINDINGS);
-            if (exact != null && IsField(exact))
+            if (exact != null)
```
Comment (byronlin13) [inline Amazon.IonObjectMapper/IonObjectSerializer.cs:381]: this can be simplified to:
```
if (exact != null && IsField(exact)) {
   return exact;
}
...
```

File: Amazon.IonObjectMapper/IonObjectSerializer.cs [hunk 6]
```diff
@@ -254,16 +246,8 @@ private bool TryDeserializeMethod(MethodInfo method, IIonReader reader, IonType
             // We deserialize whether or not we ultimately use the result because values
             // for some Ion types need to be consumed in order to advance the reader.
             deserialized = ionSerializer.Deserialize(reader, property.PropertyType, ionType);
-            
-            if (IsReadOnlyProperty(property))
```
Comment (byronlin13) [inline Amazon.IonObjectMapper/IonObjectSerializer.cs:258]: Why is this removed? Won't the deserialize logic attempt to set the value to property?

File: Amazon.IonObjectMapper/IonObjectSerializer.cs (hunk 1 above)
Comment (justing-bq) [reply Amazon.IonObjectMapper/IonObjectSerializer.cs:274]: A readonly property has a `{ get; }` accessor defined but the corresponding backing field will actually have `field.IsInitOnly = true` which could potentially result in this `IgnoreDeserializedField` method returning a false positive.

And even if you were right that the extra condition is harmless (at the moment), it's still worth separating out the IgnoreProperty and IgnoreField logic like this because we want this backing field to be treated the same as a property (however that logic is defined which could potentially change in the future).

File: Amazon.IonObjectMapper/IonObjectSerializer.cs (hunk 6 above)
Comment (justing-bq) [reply Amazon.IonObjectMapper/IonObjectSerializer.cs:258]: Thanks to the code coverage, I discovered that this condition is never satisfied (because readonly properties are previously filtered out by `GetValidProperties`).

File: Amazon.IonObjectMapper/IonObjectSerializer.cs (hunk 5 above)
Comment (justing-bq) [reply Amazon.IonObjectMapper/IonObjectSerializer.cs:381]: Nope because the following else condition is an else for `exact != null`, not for `exact != null && IsField(exact)`.

**Review by simonz-bq — APPROVED** [review]

**Review by byronlin13 — APPROVED** [review]

**Review by shuiwan — APPROVED** [review]
